src/main/java/gateways/MessageSearchFirebaseSystem.java

```diff
@@ -1,21 +1,77 @@
 package gateways;
 
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.Firestore;
+import com.google.firebase.cloud.FirestoreClient;
+import entities.Message;
+import entities.User;
 import message_search_use_case.MessageSearchData;
 import message_search_use_case.MessageSearchGateway;
 
-
-
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
 
 public class MessageSearchFirebaseSystem implements MessageSearchGateway {
+
     /**
      * Searches for messages in Google Cloud database that match text in data
      * @param data MessageSearchData object that tells method which messages to look for
      * @return list of Message objects that match data
      */
     @Override
-    public List<String> search(MessageSearchData data) {
+    public List<Message> search(MessageSearchData data) {
+        String chatId = "id" + data.getChatId(); // to know which chat Document to search in
```

Comment:
Wondering if it would be a good idea to split off some of the code in this function to a separate function then calling it here to make it more clean and readable

Reply:
That's a good idea! I'll get working on that!